File: backend/app/rbac.py
"""Role-based access control: the server-side source of truth.

Roles (per business membership):
  - owner      : full access
  - accountant : add/edit entries + view everything + reports & export; no delete,
                 no settings, no team management
  - staff      : add entries only (no lists, reports, settings, edit or delete)

Permissions are checked from the caller's *live* membership role on every request
(never trusted from the client or baked into the token, so a role change or
removal takes effect immediately).
"""

# --- Permission keys --------------------------------------------------------
ENTRY_CREATE = "entry.create"  # add income/expense/ledger/customer
ENTRY_EDIT = "entry.edit"      # edit entries/customers
ENTRY_DELETE = "entry.delete"  # delete entries/customers
DATA_VIEW = "data.view"        # dashboard, transactions, customer lists, khata
REPORTS_VIEW = "reports.view"  # view reports
REPORTS_EXPORT = "reports.export"
SETTINGS_MANAGE = "settings.manage"  # items, recurring, business config
TEAM_MANAGE = "team.manage"    # invite / change role / remove members

ALL_PERMISSIONS = {
    ENTRY_CREATE,
    ENTRY_EDIT,
    ENTRY_DELETE,
    DATA_VIEW,
    REPORTS_VIEW,
    REPORTS_EXPORT,
    SETTINGS_MANAGE,
    TEAM_MANAGE,
}

ROLES = ("owner", "accountant", "staff")

ROLE_PERMISSIONS: dict[str, set[str]] = {
    "owner": set(ALL_PERMISSIONS),
    "accountant": {
        ENTRY_CREATE,
        ENTRY_EDIT,
        DATA_VIEW,
        REPORTS_VIEW,
        REPORTS_EXPORT,
    },
    "staff": {ENTRY_CREATE},
}


def role_can(role: str, permission: str) -> bool:
    return permission in ROLE_PERMISSIONS.get(role, set())

File: src/features/auth/rbac.ts
/**
 * Client-side mirror of the backend RBAC map (backend/app/rbac.py).
 *
 * This ONLY gates the UI (hide what the server would reject). The server is the
 * real boundary — never rely on this for security, only for UX.
 */

export type Role = 'owner' | 'accountant' | 'staff';

export const PERMISSIONS = {
  ENTRY_CREATE: 'entry.create',
  ENTRY_EDIT: 'entry.edit',
  ENTRY_DELETE: 'entry.delete',
  DATA_VIEW: 'data.view',
  REPORTS_VIEW: 'reports.view',
  REPORTS_EXPORT: 'reports.export',
  SETTINGS_MANAGE: 'settings.manage',
  TEAM_MANAGE: 'team.manage',
} as const;

export type Permission = (typeof PERMISSIONS)[keyof typeof PERMISSIONS];

const ALL: Permission[] = Object.values(PERMISSIONS);

const ROLE_PERMISSIONS: Record<Role, ReadonlySet<Permission>> = {
  owner: new Set(ALL),
  accountant: new Set<Permission>([
    PERMISSIONS.ENTRY_CREATE,
    PERMISSIONS.ENTRY_EDIT,
    PERMISSIONS.DATA_VIEW,
    PERMISSIONS.REPORTS_VIEW,
    PERMISSIONS.REPORTS_EXPORT,
  ]),
  // Pure add-only: no lists, reports, settings, edit or delete.
  staff: new Set<Permission>([PERMISSIONS.ENTRY_CREATE]),
};

/** Whether a role grants a permission. Unknown/absent role → no access. */
export function roleCan(role: Role | undefined | null, permission: Permission): boolean {
  if (!role) return false;
  return ROLE_PERMISSIONS[role]?.has(permission) ?? false;
}

export const ROLE_LABEL: Record<Role, string> = {
  owner: 'Owner',
  accountant: 'Accountant',
  staff: 'Staff',
};
